### PR TITLE
Witness generation code for Keccak gates

### PR DESCRIPTION
**Pull request by querolita**
Closes https://github.com/o1-labs/proof-systems/issues/1236

### CODE REVIEW

File: kimchi/src/circuits/polynomials/keccak/mod.rs [hunk 1]
```diff
@@ -98,6 +103,82 @@ pub(crate) fn expand_word<F: PrimeField, T: ExprOps<F>>(word: u64) -> Vec<T> {
         .collect::<Vec<T>>()
 }
 
+/// Pads the message with the 10*1 rule until reaching a length that is a multiple of the rate
+pub(crate) fn pad(message: &[u8]) -> Vec<u8> {
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak/mod.rs:107]: where-is-your-5c9ae4

Comment (dannywillems) [reply kimchi/src/circuits/polynomials/keccak/mod.rs:107]: Would have been nice to have some tests with corner cases.
What about messages of 1088 bits? The documentation does not mention it. Reading the code makes me feel it adds a full block. But tests can confirm this.

File: kimchi/src/circuits/polynomials/keccak/mod.rs [hunk 2]
```diff
@@ -98,6 +103,82 @@ pub(crate) fn expand_word<F: PrimeField, T: ExprOps<F>>(word: u64) -> Vec<T> {
...
+    let mut padded = message.to_vec();
+    padded.push(0x01);
+    while padded.len() % 136 != 0 {
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak/mod.rs:110]: The code looks very imperative. I don't know which style rustaceans prefer to code in: functional or imperative?
As a functional programmer, I tend to prefer the functional style. I would then use an approach without any side effect on the vector padded. `padded.len()` has a different value every run as we push, which makes it not "functional friendly".

File: kimchi/src/circuits/polynomials/keccak/mod.rs [hunk 3]
```diff
@@ -98,6 +103,82 @@ pub(crate) fn expand_word<F: PrimeField, T: ExprOps<F>>(word: u64) -> Vec<T> {
...
+    }
+    shifts.iter().flatten().copied().collect()
+}
+
+/// From a vector of shifts, resets the underlying value returning only shift0
+pub(crate) fn reset(shifts: &[u64]) -> Vec<u64> {
+    shifts
+        .iter()
+        .copied()
+        .take(shifts.len() / 4)
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak/mod.rs:138]: There is a hypothesis on the lengths of shift here, not constrained or mentioned in the doc.

File: kimchi/src/circuits/polynomials/keccak/mod.rs [hunk 4]
```diff
@@ -98,6 +103,82 @@ pub(crate) fn expand_word<F: PrimeField, T: ExprOps<F>>(word: u64) -> Vec<T> {
...
+    let last = padded.len() - 1;
+    padded[last] += 0x80;
+    padded
+}
+
+/// From each quarter in sparse representation, it computes its 4 resets.
+/// The resulting vector contains 4 times as many elements as the input.
+/// The output is placed in the vector as [reset0, reset1, reset2, reset3]
+pub(crate) fn shift(state: &[u64]) -> Vec<u64> {
+    let mut shifts = vec![vec![]; 4];
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply kimchi/src/circuits/polynomials/keccak/mod.rs:122]: Nit: Maybe also use `QUARTER` instead of `4`?

File: kimchi/src/circuits/polynomials/keccak/mod.rs [hunk 5]
```diff
@@ -98,6 +103,82 @@ pub(crate) fn expand_word<F: PrimeField, T: ExprOps<F>>(word: u64) -> Vec<T> {
...
+    shifts
+        .iter()
+        .copied()
+        .take(shifts.len() / 4)
+        .collect::<Vec<u64>>()
+}
+
+/// From a reset0 state, obtain the corresponding 16-bit dense terms
+pub(crate) fn collapse(state: &[u64]) -> Vec<u64> {
+    let mut dense = vec![];
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak/mod.rs:144]: Is there not a equivalent of `init` to make it more `functional` friendly?

File: kimchi/src/circuits/polynomials/keccak/witness.rs [hunk 6]
```diff
@@ -31,15 +45,397 @@ fn _layout_round<F: PrimeField>() -> _Layout<F, KECCAK_COLS> {
...
+    }
+}
+
+/// Creates a witness for the Keccak hash function
+/// Input:
+/// - message: the message to be hashed
+/// Note:
+/// Requires at least one more row after the keccak gadget so that
+/// constraints can access the next row in the squeeze
+pub fn extend_keccak_witness<F: PrimeField>(witness: &mut [Vec<F>; KECCAK_COLS], message: BigUint) {
```
Comment (dannywillems) [inline kimchi/src/circuits/polynomials/keccak/witness.rs:323]: Nit: there are constraints on the field size. But I don't know if we want to ensure that in the code.

**Review by dannywillems — APPROVED** [review]
